src/typegate/src/runtimes/substantial.ts

```diff
@@ -0,0 +1,412 @@
+    this.logger = getLogger(`substantial:'${typegraphName}'`);
+    this.backend = backend;
+    this.workerManager = workerManager;
+    this.workflowRelaunchDelayMs = relaunchDelayMs;
+  }
+
+  static async init(params: RuntimeInitParams): Promise<Runtime> {
+    logger.info("initializing SubstantialRuntime");
+    logger.debug(`init params: ${JSON.stringify(params)}`);
+    const {
+      typegraph: tg,
+      args,
+      materializers,
+      secretManager,
+      typegate,
+    } = params;
+
+    const secrets: Record<string, string> = {};
+
+    for (const m of materializers) {
+      for (const secretName of (m.data.secrets as []) ?? []) {
+        secrets[secretName] = secretManager.secretOrFail(secretName);
+      }
+    }
+
+    const tgName = TypeGraph.formatName(tg);
+    const backend = (args as any)!.backend as native.Backend;
+    const workerManager = WorkerManager.getInstance();
+
+    const instance = new SubstantialRuntime(
+      tgName,
+      backend,
+      workerManager,
+      typegate.config.base.substantial_relaunch_ms
+    );
+
+    await instance.#prepareWorkflowFiles(
+      tg.meta.artifacts,
+      materializers,
+      typegate
+    );
+
+    return instance;
+  }
+
+  deinit(): Promise<void> {
+    logger.info("deinitializing SubstantialRuntime");
+    this.workerManager.destroyAllWorkers();
+    return Promise.resolve();
+  }
+
+  materialize(
+    stage: ComputeStage,
+    _waitlist: ComputeStage[],
+    _verbose: boolean
+  ): ComputeStage[] {
+    if (stage.props.node === "__typename") {
+      return [
+        stage.withResolver(() => {
+          const { parent: parentStage } = stage.props;
+          if (parentStage != null) {
+            return parentStage.props.outType.title;
+          }
+          switch (stage.props.operationType) {
+            case ast.OperationTypeNode.QUERY:
+              return "Query";
+            case ast.OperationTypeNode.MUTATION:
+              return "Mutation";
+            default:
+              throw new Error(
+                `Unsupported operation type '${stage.props.operationType}'`
+              );
+          }
+        }),
+      ];
+    }
+
+    if (stage.props.outType.config?.__namespace) {
+      return [stage.withResolver(() => ({}))];
+    }
+
+    if (stage.props.materializer != null) {
+      return [stage.withResolver(this.#delegate(stage.props.materializer))];
+    }
+
+    return [
+      stage.withResolver(({ _: { parent } }) => {
+        if (stage.props.parent == null) {
+          // namespace
+          return {};
+        }
+        const resolver = parent[stage.props.node];
+        return typeof resolver === "function" ? resolver() : resolver;
+      }),
+    ];
+  }
+
+  #delegate(mat: TypeMaterializer): Resolver {
+    const matName = mat.name;
+    const data = mat?.data ?? {};
+    const workflowName = data.name as string;
+
+    switch (matName) {
+      case "start":
+        return this.#startResolver(workflowName);
```

Comment:
The current implementation immediately starts the workflow. The poc implementation would put this in a schedule timeline registry-ish, and then an infinite loop would poll for the scheduled workflows in time order but it's all done automatically with workers.
Caveat being that `Start` information is lost if the typegate dies (we lose `resume` feature), so I am currently considering of adding it unless you have suggestions.

Reply:
Also marking this as a resolved question, workflow runs will be resumed if the typegate dies and restarted.